common/mqtt_client/src/lib.rs

```diff
@@ -239,16 +184,22 @@ impl Client {
                         QoS::ExactlyOnce => {
                             // Do not annouce the incoming publish message immediatly in case
                             // of QoS=2. Wait for the `PubRel`.
-                            //
-                            // TODO: Avoid buffering the message here by
-                            // moving that part of the code into the `rumqttc` library.
                             let _ = pending.insert(msg.pkid, msg.into());
                         }
                     }
                 }
 
+                Ok(Event::Incoming(Packet::PubAck(_))) => {
+                    pending_count
+                        .as_ref()
+                        .fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
+                }
+
                 Ok(Event::Incoming(Packet::PubRel(pubrel))) => {
                     if let Some(msg) = pending.remove(&pubrel.pkid) {
+                        pending_count
+                            .as_ref()
+                            .fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
```

Comment:
This must be removed. Indeed, the `PubRel` message is related to received messages, while the `pending_count` is related to published messages.

```suggestion
```